AllClasses and AllClassWithGeneric based on GDB

jeffmaury opened the pull request:
Fixes #4

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

Fixes #

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

-
-
-

<!--
If this change has user-visible impact, follow the instructions below.
Examples include:

- :gift: Add new feature
- :bug: Fix bug
- :broom: Update or clean up current behavior
- :wastebasket: Remove feature or internal logic

Otherwise delete the rest of this template.
-->

jeffmaury committed:
AllClasses and AllClassWithGeneric based on GDB

Fixes #4

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

aavarghese commented on src/main/java/jdwp/Translator.java:
```diff
@@ -258,6 +362,36 @@ public static LocationImpl locationLookup(String func, int line) {
+		for(SymbolFileInfo symbolFile : response.getSymbolFiles()) {
+			for(Symbols symbol : symbolFile.getSymbols()) {
+				var index = symbol.getName().indexOf("::");
+				if (index != (-1)) {
```
@jeffmaury What do you think about instead of generating the full output for `-symbol-info-functions` and filtering out with an `if` check here, we can call the GDB MI method at [loadSymbols()](https://github.com/nativejdb/nativejdb/pull/71/files#diff-818cd7a448edb115cc7340441d8bf5a1d33760422cb675a2058b7073f0a72f3bR113) with `createMiSymbolInfoFunctions("", "::", 0, false);`?

Since the output is huge, it takes a long time to generate and respond...

jeffmaury committed:
Use :: as a filter for --symbol-info-functions to lower output

Suggested by @aavarghese

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Fix failing unit tests

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Remove scaffolfing from VirtualMachine.Version

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Introduce ReferenceTypes container to better handle lookups

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
GDB based implementation of:

- ClassType/Superclass
- EventRequest/Set
- Method/LineTable
- ReferenceType/Methods
- ReferenceType/SourceFile
- ReferenceType/SignatureWithGenerics
- ReferenceType/MethodsWithGenerics
- StackFrame/GetValues
- ThreadReference/Frames
- VirtualMachine/ClassesBySignature

Added a TypeEnricher to get information from source files about types/methods that can't be extracted from GDB

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
GDB based implementation of:

- ObjectReference/ReferenceType
- ReferenceType/Interfaces
- ReferenceType/Fields
- StackFrame/GetValues


Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Return Object id in GetValues (from String,....)

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Implement StepIn/StepOver

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Reject Method/Bytecodes

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Fix StepOut

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Fix wrong tag for ReferenceType

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

aavarghese commented:
Running this branch as-is throws this CNFE err when running nativejdb on Docker with any example:
```
Waiting for debugger on: b64602176b9c:8082
Exception in thread "main" java.lang.NoClassDefFoundError: com/github/javaparser/symbolsolver/model/resolution/TypeSolver
	at jdwp.model.TypeEnricherFactory.lambda$getTypeEnricher$0(TypeEnricherFactory.java:14)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1134)
	at jdwp.model.TypeEnricherFactory.getTypeEnricher(TypeEnricherFactory.java:14)
	at jdwp.model.ReferenceTypes.<init>(ReferenceTypes.java:16)
	at jdwp.GDBControl.<init>(GDBControl.java:69)
	at jdwp.JDWPProxy.reply(JDWPProxy.java:88)
	at jdwp.JDWPServer.main(JDWPServer.java:83)
Caused by: java.lang.ClassNotFoundException: com.github.javaparser.symbolsolver.model.resolution.TypeSolver
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 7 more
	```

jeffmaury commented:
> Running this branch as-is throws this CNFE err when running nativejdb on Docker with any example:
> 
> ```
> Waiting for debugger on: b64602176b9c:8082
> Exception in thread "main" java.lang.NoClassDefFoundError: com/github/javaparser/symbolsolver/model/resolution/TypeSolver
> 	at jdwp.model.TypeEnricherFactory.lambda$getTypeEnricher$0(TypeEnricherFactory.java:14)
> 	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1134)
> 	at jdwp.model.TypeEnricherFactory.getTypeEnricher(TypeEnricherFactory.java:14)
> 	at jdwp.model.ReferenceTypes.<init>(ReferenceTypes.java:16)
> 	at jdwp.GDBControl.<init>(GDBControl.java:69)
> 	at jdwp.JDWPProxy.reply(JDWPProxy.java:88)
> 	at jdwp.JDWPServer.main(JDWPServer.java:83)
> Caused by: java.lang.ClassNotFoundException: com.github.javaparser.symbolsolver.model.resolution.TypeSolver
> 	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
> 	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
> 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
> 	... 7 more
> 	```
> ```

NativeJDB now has a dependency on a third library (Java Parser) so we should either update the scripts or publish a NativeJDB jar with dependencies included

jeffmaury committed:
Add uber JAR to fix examples

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Always return system class loader

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Implement ArrayReference

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Handle String values

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Handle compressed references

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
WIP

jeffmaury committed:
Restore $asm variable

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Remove scaffolding VM

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Add fields support

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

jeffmaury committed:
Restore CLASS_PREPARE events

Signed-off-by: Jeff MAURY <jmaury@redhat.com>
